changelogs/unreleased/gh-5272-bootstrap-strategy-auto.md

```diff
@@ -2,7 +2,18 @@
 
 * **[Breaking change]** Introduce new configuration option `bootstrap_strategy`.
   The default value of this option - "auto" - brings new replica behaviour on
-  replica set bootstrap, recovery and on replication reconfiguration. One of the
-  notable changes is that `replication_connect_quorum` option is not taken into
-  account with `bootstrap_strategy` = "auto". If you wish to return to the old
-  behaviour, you may use `bootstrap_strategy` = "legacy" (deprecated) (gh-5272).
+  replica set bootstrap, recovery and on replication reconfiguration. The user
+  may set the option to "legacy" to preserve old behaviour.
+
+  The new option value - "auto" - will be in effect only when no value for
+  `replication_connect_quorum` is passed. If the user passes any value of
+  `replication_connect_quorum`, `bootstrap_strategy` is automatically set to
+  "legacy" for backward compatibility reasons.
+
+  Examples of changed behaviour (when neither `bootstrap_strategy` nor
+  `replication_connect_quorum` are passed explicitly) include:
+  1) bootstrapping a new replica in an existing cluster will only succeed if all
+  the cluster members are listed in replica's `box.cfg.replication`.
+  2) when reconfiguring `box.cfg.replication`, the node will not stay orphan
+  even if some of the remote peers are unavailable. The node will transition
+  from orphan to running as soon as all the alive peers are synced (gh-5272).
```

Comment:
A few questions on the new changelog entry.

1. We agreed to split breaking and non-breaking changes. This entry is marked as breaking at whole. I would split it into two entries (we can add several entries within one file).
2. Are the examples in the last paragraph present exhaustive list of breaking changes? If so, it worth to rephrase it not as examples, but as the list of breaking changes.

Reply:
1. Ok
2. Yes, the list is exhaustive.

Now the changelog looks like:
```markdown
## feature/replication

* Introduce new configuration option `bootstrap_strategy`. The default value of
  this option - "auto" - brings new replica behaviour on replica set bootstrap,
  recovery and on replication reconfiguration. The user may set the option to
  "legacy" to preserve old behaviour.

  The new option value - "auto" - will be in effect only when no value for
  `replication_connect_quorum` is passed. If the user passes any value of
  `replication_connect_quorum`, `bootstrap_strategy` is automatically set to
  "legacy" for backward compatibility reasons (gh-5272).

* **[Breaking change]** If options `bootstrap_strategy` and
  `replication_connect_quorum` are not passed explicitly, the following
  behaviour will change:
  1) bootstrapping a new replica in an existing cluster will only succeed if all
  the cluster members are listed in replica's `box.cfg.replication`.
  2) when reconfiguring `box.cfg.replication`, the node will not stay orphan
  even if some of the remote peers are unavailable. The node will transition
  from orphan to running as soon as all the alive peers are synced (gh-5272).
```

Reply:
**upd**: after our verbal discussion, changed the changelog:
```markdown
## feature/replication

* Introduce new configuration option `bootstrap_strategy`. The default value of
  this option - "auto" - brings new replica behaviour on replica set bootstrap,
  recovery and on replication reconfiguration. The user may set the option to
  "legacy" to fully return to old behaviour.

  The new option value - "auto" - will be in effect only when no value for
  `replication_connect_quorum` is passed. If the user passes any value of
  `replication_connect_quorum`, `bootstrap_strategy` is automatically set to
  "legacy" for backward compatibility reasons.

  Note, that if you leave the options untouched (that is, `bootstrap_strategy`
  defaults to "auto"), the following behaviour will noticeably change: during
  the recovery from local files and during replication reconfiguration the
  replica will not stay "orphan" even if some of the remote peers listed in
  `box.cfg.replication` are unavailable. The node will transition from "orphan"
  to "running" as soon as it syncs with all the reachable peers (unlike before,
  when not setting `replication_connect_quorum` would mean that each remote peer
  connection is necessary). Note, the node will try to establish connections
  with all the peers listed in `box.cfg.replication` for a period of
  `replication_connect_timeout` seconds (gh-5272).

* **[Breaking change]** If neither of the configuration options
  `bootstrap_strategy` and `replication_connect_quorum` is passed explicitly, or
  if `bootstrap_strategy` is set to "auto", bootstrapping a new replica in an
  existing replica set will only succeed if all the replica set members are
  listed in replica's `box.cfg.replication`. For example, when joining a fresh
  replica to a replica set of 3 nodes, all 3 node URIs must be present in
  replica's `box.cfg.replication` parameter. When joining 2 new replicas to a
  single master, both replicas must have each other's URIs (alongside with
  master's URI) in their `box.cfg.replication` (gh-5272).
```

Reply:
And one more iteration:
```markdown
## feature/replication

* Introduce new configuration option `bootstrap_strategy`. The default value of
  this option - "auto" - brings new replica behaviour on replica set bootstrap,
  replica join to an existing replica set, recovery and replication
  reconfiguration. The user may set the option to "legacy" to fully return to
  old behaviour.

  The new option value - "auto" - will be in effect only when no value for
  `replication_connect_quorum` is passed. If the user passes any value of
  `replication_connect_quorum`, `bootstrap_strategy` is automatically set to
  "legacy" for backward compatibility reasons.

  Note, that if you leave the options untouched (that is, `bootstrap_strategy`
  defaults to "auto"), the following behaviour will noticeably change: during
  the recovery from local files and during replication reconfiguration
  `box.cfg{replication = ...}` will not fail even if some (or all) of the remote
  peers listed in `box.cfg.replication` are unavailable. Instead the node will
  try to establish a connection to everyone for a period of
  `replication_connect_timeout` and then transition to
  `box.info.status == "running"` as soon as it syncs with all the reached peers.
  (gh-5272).

* **[Breaking change]** If neither of the configuration options
  `bootstrap_strategy` and `replication_connect_quorum` is passed explicitly, or
  if `bootstrap_strategy` is set to "auto", bootstrapping a new replica in an
  existing replica set will only succeed if all the replica set members are
  listed in replica's `box.cfg.replication`. For example, when joining a fresh
  replica to a replica set of 3 nodes, all 3 node URIs must be present in
  replica's `box.cfg.replication` parameter. When joining 2 new replicas to a
  single master, both replicas must have each other's URIs (alongside with
  master's URI) in their `box.cfg.replication` (gh-5272).
```

Reply:
Thank you for the updates! I think we reached a good balance between verbosity and simplicity.